### PR TITLE
refactor: add logger test and don't log to console on production

### PR DESCRIPTION
**Pull request by theahmadzai**
Added a logger test for validateInput internal method with the help of rewire, the logger must have log function when no logger is provided in this case console.log.
Added a check not to log to console on production
Removed an unnecessary check, checking if opts.logger is defined then logger.log otherwise log to console, while the logger is always returned from the validateInput internal method.

### CODE REVIEW

**Review by vhashimotoo — DISMISSED** [review]
Thanks for your contribution, could you please describe why need to disable logging on production, this seems usable for using inside the error reporting systems.

Also, that seems like a breaking change since you can't change this behaviour to old, I correctly understand that?

**Comment by theahmadzai** [comment]
Thanks for your remarks, No not disable logging for production if the user has not specified any logger it defaults to console that's why I'm disabling logging to "console" on production if no other logger is provided because the console is useless on production while we have sentry and other services to catch errors. Still, I can change it back to how it was.

**Comment by HashimotoYT** [comment]
However, I do not see any problem with using `console` as default logger, they normally work with error reporting systems.
As the best solution for this, I think we can just add the ability to provide `boolean` value to `logger` parameter with `false` value to disable the logging. This helps to just not create breaking changes and allows us to manually enable or disable logging.

**Comment by theahmadzai** [comment]
Sounds great, but what about nullable value, If the logger is provided null then don't log irrespective of the env, otherwise, default to console, this way the developer can explicitly set the logger to null if he doesn't want to log in production.

**Comment by theahmadzai** [comment]
When no logger is provided console is the default, on falsy logger value it doesn't log

File: index.d.ts [hunk 1]
```diff
@@ -23,7 +23,7 @@ export function updater(
          *                        Defaults to `console`. See electron-log, a module
          *                        that aggregates logs from main and renderer processes into a single file.
          */
-        logger?: ILogger;
+        logger?: ILogger | null | undefined | false;
```
Comment (HashimotoYT) [inline index.d.ts:26]: TypeScript automatically give `undefined` type for optional variables.

```suggestion
        logger?: ILogger | null | false;
```

**Review by HashimotoYT — COMMENTED** [review]
Seems fine, can you please rebase your PR to trigger the CI for verify tests.

**Review by theahmadzai — COMMENTED** [review]
Resolved

**Review by erickzhao — CHANGES_REQUESTED** [review]
Could we update the [documentation in the README](https://github.com/electron/update-electron-app#updateoptions) to reflect the new console logging change?

File: readme.md [hunk 2]
```diff
@@ -56,8 +56,8 @@ Options:
 - `repo` String (optional) - A GitHub repository in the format `owner/repo`. Defaults to your `package.json`'s `"repository"` field
 - `host` String (optional) - Defaults to `https://update.electronjs.org`
 - `updateInterval` String (optional) - How frequently to check for updates. Defaults to `10 minutes`. Minimum allowed interval is `5 minutes`.
-- `logger` Object (optional) - A custom logger object that defines a `log` function. Defaults to `console`. See [electron-log](https://github.com/megahertz/electron-log), a module that aggregates logs from main and renderer processes into a single file.
-- `notifyUser` Boolean (optional) - Defaults to `true`.  When enabled the user will be
+- `logger` Object | null | false (optional) - A custom logger object that defines a `log` function. Defaults to `console`. See [electron-log](https://github.com/megahertz/electron-log), a module that aggregates logs from main and renderer processes into a single file. To disable logging provide a null or any falsy value.
```
Comment (erickzhao) [inline readme.md:59]:
```suggestion
- `logger` Object | null | false (optional) - A custom logger object that defines a `log` function. Defaults to `console`. See [electron-log](https://github.com/megahertz/electron-log), a module that aggregates logs from main and renderer processes into a single file. To disable logging, omit this option or provide any falsy value.
```

Comment (theahmadzai) [reply readme.md:59]: By omitting this option it defaults to console, only null undefined or false can disable it.

File: index.js [hunk 3]
```diff
@@ -24,7 +24,7 @@ module.exports = function updater (opts = {}) {
   // don't attempt to update during development
   if (isDev) {
     const message = 'update-electron-app config looks good; aborting updates since app is in development mode'
-    opts.logger ? opts.logger.log(message) : console.log(message)
```
Comment (erickzhao) [inline index.js:27]: What's the purpose of removing the `console.log` here if it's gated in dev mode?

Comment (theahmadzai) [reply index.js:27]: Good catch, previously the opts.logger would always return a logger that's why I refactored it, now when we added the option to disable logger, this line will throw error complaining log doesn't exist on the object if someone set the logger to false, we should restore this line.

Comment (theahmadzai) [reply index.js:27]: maybe
```javascript
opts.logger && opts.logger.log(message)
```

**Comment by theahmadzai** [comment]
I think we can close this, no longer relevant.

**Pull request closed without merging**